### PR TITLE
Xfconf Tutorial

### PR DESCRIPTION
**Pull request by Eiskasten**
This PR adds a tutorial for Xfconf as mentioned in issue #280.

## Basics

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add them to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - add a line in `doc/news/_preparation_next_release.md`
  - reformat the code with `scripts/dev/reformat-all`
  - make all unit tests pass
  - fix all memleaks
  - fix the CI itself (or rebase if already fixed)
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if everything is done** and no further pushes are planned by you.

### CODE REVIEW

**Comment by Eiskasten** [comment]
jenkins build libelektra please

File: doc/tutorials/xfconf.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+**Warning**: The following operations will cause permanent changes to your system, please handle with care.
+
+Thunar should now be ready to be configured by Elektra.
+For example, the following command can be used to use single clicks instead of double clicks.
+
+```shell
+# Setting the value to `false` beforehand
+xfconf-query -c thunar -p /misc-single-click -s false -t bool -n
+
+kdb set /sw/xfce4/thunar/misc-single-click TRUE
```
Comment (hannes99) [inline doc/tutorials/xfconf.md:32]: Is `dir:` where this should end up? Maybe `user:` or spec might make more sense, no?

**Review by hannes99 — APPROVED** [review]
Maybe a short section on when what namespace makes sense. Other than that LGTM!

File: doc/tutorials/xfconf.md (hunk 1 above)
Comment (Eiskasten) [reply doc/tutorials/xfconf.md:32]: Thank you for your review! You are right, the namespace which makes most sense to me are `user:` and `system:`. I just have overseen, that the default seemed to change.

**Review by tucek — APPROVED** [review]
LGTM!

**Comment by Eiskasten** [comment]
jenkins build libelektra please

File: doc/tutorials/xfconf.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+```
+
+Note that a system upgrade may reset the system library replacement.
+
+All applications that normally use Xfconf will now use Elektra instead.
+This can be verified using both `xfconf-query` and `kdb`.
+
+```shell
+# Set an Xfconf property with Elektra
+kdb set system:/sw/org/xfce/xfconf/test/hello world
```
Comment (markus2330) [inline doc/tutorials/xfconf.md:84]: `kdb: symbol lookup error: /usr/lib/libxfconf-elektra.so: undefined symbol: g_log`

File: src/plugins/xfconf/README.md [hunk 3]
```diff
@@ -2,9 +2,9 @@
 - infos/author = Richard Stöckl <e11908080@student.tuwien.ac.at>
 - infos/licence = BSD
 - infos/needs =
-- infos/provides = storage/xfconf
+- infos/provides = storage/xml
```
Comment (markus2330) [inline src/plugins/xfconf/README.md:5]: This should be in a different PR to merge it first.

File: doc/tutorials/xfconf.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+Xfconf uses the glib2 type system.
+This system is not fully compatible with Elektra.
+However, all values are stored in Elektra in their string representation when used as a drop-in replacement.
+In addition, the name of the glib2 type is stored in Elektra as a metakey.
+This makes Xfconf's type system completely independent of Elektra, at the cost of memory and computational resources.
+
+The following table shows a comparison between the different type systems.
+
+| Xfconf   | Elektra                    | Note                                                                                         |
```
Comment (markus2330) [inline doc/tutorials/xfconf.md:131]: Please also a xfce4 example per type.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Tutorial is great and easy to follow but doesn't provide enough material to be actually able to reconfigure xfce4 and unfortunately did not work on Debian bullseye.

**Comment by markus2330** [comment]
Btw. I think also restart of xfconfd is needed so that some parts of the tutorial actually can work? (Logout/Login does not restart xfconfd for me, and without restart it cannot pick up a different library).

File: doc/tutorials/xfconf.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+**Warning**: The following operations will cause permanent changes to your system, please handle with care.
+
+Thunar should now be ready to be configured by Elektra.
+For example, the following command can be used to use single clicks instead of double clicks.
+
+```shell
+# Setting the value to `false` beforehand
+xfconf-query -c thunar -p /misc-single-click -s false -t bool -n
+
+kdb set system:/sw/xfce4/thunar/misc-single-click TRUE
```
[the remark this answers is not in the document]
Comment (Eiskasten) [reply doc/tutorials/xfconf.md]: #4947 fixed this. Since I rebased this PR after you tried it, that should work now.

File: doc/tutorials/xfconf.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+The Xfconf binding in Elektra implements the Xfconf API in such a way that it can be used as a drop-in replacement.
+
+### Setup
+
+To start, the Xfconf library must be replaced with the Xfconf binding.
+Replacing the system libraries may require root privileges.
+
+```shell
+# Use Elektra instead of Xfconf system-wide
+kdb xfconf-system-lib-replace
```
[the remark this answers is not in the document]
Comment (Eiskasten) [reply doc/tutorials/xfconf.md:68]: #4960 Should solve this

File: doc/tutorials/xfconf.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+# Use Elektra instead of Xfconf only for the current user
+kdb xfconf-user-lib replace
+
+# When replacing only the users library, either restart the session or run
+source ~/.xprofile
```
[the remark this answers is not in the document]
Comment (Eiskasten) [reply doc/tutorials/xfconf.md]: #4960 should solve this

File: doc/tutorials/xfconf.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+All applications that normally use Xfconf will now use Elektra instead.
+This can be verified using both `xfconf-query` and `kdb`.
+
+```shell
+# Set an Xfconf property with Elektra
+kdb set system:/sw/org/xfce/xfconf/test/hello world
+#> Create a new key system:/sw/org/xfce/xfconf/test/hello with string "world"
+
+# Verify it with Xfconf
+xfconf-query -c test -p hello
```
[the remark this answers is not in the document]
Comment (Eiskasten) [reply doc/tutorials/xfconf.md]: This is solved by 5d83e517863f0f47d0d41e0475109073cbcc8604

File: doc/tutorials/xfconf.md [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+```shell
+kdb xfconf-populate
+```
+
+Once this is done, an Xfce session can be started either from the login manager or manually.
+From this point on, the Xfce desktop will read and write to the Elektra database.
+However, Xfce also uses some properties that are not stored in Xfconf.
+These properties will not work properly.
+An example of such a property is the Gtk theme.
+
```
[the remark this answers is not in the document]
Comment (Eiskasten) [reply doc/tutorials/xfconf.md:104]: I changed it a bit in abb652380717583a1a71e9a75ab3e0cb3f7804ec which should work now

File: doc/tutorials/xfconf.md [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+To start, the Xfconf library must be replaced with the Xfconf binding.
+Replacing the system libraries may require root privileges.
+
+```shell
+# Use Elektra instead of Xfconf system-wide
+kdb xfconf-system-lib-replace
+
+# Use Elektra instead of Xfconf only for the current user
+kdb xfconf-user-lib replace
```
[the remark this answers is not in the document]
Comment (Eiskasten) [reply doc/tutorials/xfconf.md]: Done in 8486a892fc885f5fb0bd27cf30447c207bba9a54

File: src/plugins/xfconf/README.md (hunk 3 above)
Comment (Eiskasten) [reply src/plugins/xfconf/README.md]: This was actually an accident. I committed it in two different PRs, the other one was #4947. Since I rebased this PR with the master, this PR will not change anything regarding this.

File: doc/tutorials/xfconf.md (hunk 4 above)
Comment (Eiskasten) [reply doc/tutorials/xfconf.md]: Xfce actually only uses a few of them. I added examples to the types I have found in abb652380717583a1a71e9a75ab3e0cb3f7804ec.

**Comment by Eiskasten** [comment]
> Tutorial is great and easy to follow but doesn't provide enough material to be actually able to reconfigure xfce4 and unfortunately did not work on Debian bullseye.

Thank you for the review. I provided corrections for the requested things.

File: doc/tutorials/xfconf.md (hunk 2 above)
Comment (markus2330) [reply doc/tutorials/xfconf.md:81]: Looks like some dependency is missing.

Did you test also on Debian bullseye? In general some additional testing would be great.

**Comment by Eiskasten** [comment]
I have tested and fixed it for debian now and it works with the buster image within the repository. However, please make sure to use #4960 since this PR contains fixes which are mandatory to debian.

**Comment by Eiskasten** [comment]
jenkins build libelektra please

**Comment by markus2330** [comment]
jenkins build libelektra please
